src/diffusers/models/attention_processor.py

```diff
@@ -360,15 +360,22 @@ def prepare_attention_mask(self, attention_mask, target_length, batch_size=None,
         if attention_mask is None:
             return attention_mask
 
-        if attention_mask.shape[-1] != target_length:
+        current_length: int = attention_mask.shape[-1]
+        if current_length > target_length:
+            # we *could* trim the mask with:
+            #   attention_mask = attention_mask[:,:target_length]
+            # but this is weird enough that it's more likely to be a mistake than a shortcut
+            raise ValueError(f"mask's length ({current_length}) exceeds the sequence length ({target_length}).")
+        elif current_length < target_length:
             if attention_mask.device.type == "mps":
                 # HACK: MPS: Does not support padding by greater than dimension of input tensor.
                 # Instead, we can manually construct the padding tensor.
                 padding_shape = (attention_mask.shape[0], attention_mask.shape[1], target_length)
                 padding = torch.zeros(padding_shape, dtype=attention_mask.dtype, device=attention_mask.device)
                 attention_mask = torch.cat([attention_mask, padding], dim=2)
             else:
-                attention_mask = F.pad(attention_mask, (0, target_length), value=0.0)
+                remaining_length: int = target_length - current_length
+                attention_mask = F.pad(attention_mask, (0, remaining_length), value=0.0)
```

Comment:
This is a change in logic, no? Could I have a quick explanation on how it's backwards compatible?

Reply:
intentionally not backwards compatible.

when `current_length=3` and `target_length=4`,

original behaviour was "pad sequence to length=7":  
<img width="471" alt="image" src="https://github.com/huggingface/diffusers/assets/6141784/eb896e5c-93af-4cad-8518-f545d6b9c81e">

I didn't find this very useful:  
```
>       hidden_states = F.scaled_dot_product_attention(
            query, key, value, attn_mask=attention_mask, dropout_p=0.0, is_causal=False
        )
E       RuntimeError: The size of tensor a (4) must match the size of tensor b (7) at non-singleton dimension 3
```

I thought it would be more useful to pad the sequence length to 4:  
<img width="408" alt="image" src="https://github.com/huggingface/diffusers/assets/6141784/f27eeafd-eb6d-4ca1-a32b-f70cb2756c2c">

this throws fewer errors.

I have now added a test for this behaviour: `tests/models/test_models_unet_2d_condition.py#test_model_xattn_padding`

Reply:
Unfortunately, this padding logic is used by the unclip pipeline which uses the added kv attention processor block. If you remove the conditional [here](https://github.com/huggingface/diffusers/pull/2634/files#r1195848796) so the attention mask is passed in and run the unclip pipeline, it'll error. 

If we want to change this logic, I think it will have to be in a separate PR that also changes how the added kv attention processor block expects its attention mask to be padded. Separately, if we do change it, we'll have to change the MPS branch so it mirrors the update.

Reply:
@williamberman I've changed it back to padding by `target_length`. I've kept the test in (marked as "skip"), and added comments explaining how to restore the functionality in future.

I think this is no big loss for for stable-diffusion's cross-attn mask use-case. if somebody passes in a mask that's too few tokens… I don't really mind if it goes bang. think it's pretty weird to try and "repair" the mask anyway.